app/Jobs/FetchNativeBalances.php

```diff
@@ -6,52 +6,102 @@
 
 use App\Jobs\Traits\RecoversFromProviderErrors;
 use App\Jobs\Traits\WithWeb3DataProvider;
+use App\Models\Balance;
 use App\Models\Network;
 use App\Models\Wallet;
+use App\Support\Facades\Moralis;
 use Carbon\Carbon;
 use DateTime;
 use Illuminate\Bus\Queueable;
 use Illuminate\Contracts\Queue\ShouldBeUnique;
 use Illuminate\Contracts\Queue\ShouldQueue;
 use Illuminate\Foundation\Bus\Dispatchable;
 use Illuminate\Queue\InteractsWithQueue;
-use Illuminate\Queue\SerializesModels;
+use Illuminate\Support\Collection;
 use Illuminate\Support\Facades\DB;
+use Illuminate\Support\Facades\Log;
+use Illuminate\Support\Str;
 
 class FetchNativeBalances implements ShouldBeUnique, ShouldQueue
 {
-    use Dispatchable, InteractsWithQueue, Queueable, RecoversFromProviderErrors, SerializesModels, WithWeb3DataProvider;
+    use Dispatchable, InteractsWithQueue, Queueable, RecoversFromProviderErrors, WithWeb3DataProvider;
 
+    /**
+     * @var Collection<int, Wallet>
+     */
+    public Collection $wallets;
+
+    /**
+     * @param  Collection<int, Wallet>  $wallets
+     */
     public function __construct(
-        public Wallet $wallet,
+        Collection|Wallet $wallets,
         public Network $network,
     ) {
+        $this->wallets = $wallets instanceof Wallet ? collect([$wallets]) : $wallets;
     }
 
     public function handle(): void
     {
+        Log::info('Processing FetchNativeBalances job', [
+            'network_id' => $this->network->id,
+            'wallet_ids' => $this->wallets->pluck('id')->toArray(),
+        ]);
+
         $nativeToken = $this->network->nativeToken()->firstOrFail();
 
-        $provider = $this->getWeb3DataProvider();
-        $balance = $provider->getNativeBalance($this->wallet, $this->network);
+        $addresses = $this->wallets->pluck('address')->toArray();
+
+        $balances = Moralis::getNativeBalances($addresses, $this->network);
+
+        $walletsToUpdate = collect();
+
+        $balancesToInsert = $balances->map(function ($walletBalance) use ($nativeToken, $walletsToUpdate) {
+            $address = Str::lower($walletBalance->address);
+
+            $wallet = $this->wallets->first(fn ($wallet) => $address === Str::lower($wallet->address));
+
+            $walletsToUpdate->push($wallet);
```

Comment:
Won't this `$walletsToUpdate` variable always match `$this->wallets`? You don't need to hand-pick which wallets are going to be updated, you can just do `$this->wallets->map()` to save their `native_balances_fetched_at`

Reply:
There is no guarantee that API will return balances for all provided wallets, that is why I chose to be on the safe side. If you are 100% sure, I can remove that part.

Reply:
Even with randomly made up addresses, they still return 0 for balance (see image). Doesn't really matter as it will query 25 wallets, whatever you feel like

![Screenshot 2023-09-15 at 12 05 25](https://github.com/ArdentHQ/dashbrd/assets/6536260/8dadf412-6f59-474b-9561-85dcfaa1a7c5)

Reply:
Thanks for checking. Changed it as you suggested